File: bindings/android/lib/src/test/kotlin/org/universalpaymentqr/UniversalPaymentQRTest.kt
package org.universalpaymentqr

import org.json.JSONObject
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * Runs on the host JVM against a host-native build of the JNI crate - see
 * `build.gradle.kts`'s `java.library.path` wiring and README.md "Local unit tests" for how that
 * build gets produced. The real Android-targeted `.so` files under `src/main/jniLibs` are only
 * exercised on an actual device/emulator.
 */
class UniversalPaymentQRTest {
    @Test
    fun `parses UPI without floating point drift`() {
        val result = UniversalPaymentQR.parse("upi://pay?pa=merchant%40bank&pn=Example&am=499.00&cu=INR")
        assertTrue(result.getBoolean("recognized"))
        assertEquals("upi", result.getString("scheme"))
        assertEquals("499.00", result.getString("amount"))
    }

    @Test
    fun `recognizes a bare bitcoin address with no uri scheme`() {
        val result = UniversalPaymentQR.parse("1BoatSLRHtKNngkdXEeobR76b53LETtpyT")
        assertTrue(result.getBoolean("recognized"))
        assertEquals("bitcoin", result.getString("scheme"))
    }

    @Test
    fun `non-payment QR is recognized but flagged, not left fully unrecognized`() {
        val result = UniversalPaymentQR.parse("https://example.com/docs")
        assertTrue(result.getBoolean("recognized"))
        assertEquals("url", result.getString("scheme"))
        assertFalse(result.getBoolean("supported"))
        assertEquals("NOT_PAYMENT_QR", result.getJSONObject("support").getString("reason"))
    }

    @Test
    fun `disabled scheme stays recognized but unsupported`() {
        val scanner = createScanner(schemes = mapOf("solana_pay" to false))
        val result = scanner.scan("solana:9xQeWvG816bUx9EPfEZi4q3G44E2sA6v7a6D9k5GmRse?amount=1.25")
        assertTrue(result.getBoolean("recognized"))
        assertFalse(result.getBoolean("supported"))
        assertEquals("SCHEME_DISABLED", result.getJSONObject("support").getString("reason"))
    }

    @Test
    fun `preset scopes to india only`() {
        val scanner = createScanner(preset = "india")
        val upi = scanner.scan("upi://pay?pa=merchant%40bank&am=1&cu=INR")
        assertTrue(upi.getBoolean("supported"))
        val bitcoin = scanner.scan("1BoatSLRHtKNngkdXEeobR76b53LETtpyT")
        assertTrue(bitcoin.getBoolean("recognized"))
        assertFalse(bitcoin.getBoolean("supported"))
    }

    @Test
    fun `raw policy JSONObject overload works directly`() {
        val scanner = createScanner(JSONObject().put("preset", "crypto-only"))
        val result = scanner.scan("1BoatSLRHtKNngkdXEeobR76b53LETtpyT")
        assertTrue(result.getBoolean("supported"))
    }

    @Test
    fun `capabilities lists every registered scheme`() {
        val ids = (0 until UniversalPaymentQR.capabilities().length())
            .map { UniversalPaymentQR.capabilities().getJSONObject(it).getString("id") }
            .toSet()
        assertTrue(ids.contains("upi"))
        assertTrue(ids.contains("pix"))
        assertTrue(ids.size > 30)
    }

    @Test
    fun `supported schemes excludes experimental`() {
        val schemes = UniversalPaymentQR.supportedSchemes()
        for (i in 0 until schemes.length()) {
            assertFalse(schemes.getJSONObject(i).optString("maturity") == "experimental")
        }
    }

    @Test
    fun `version is non-empty`() {
        assertTrue(UniversalPaymentQR.version.isNotEmpty())
    }
}

File: bindings/flutter/test/universal_payment_qr_test.dart
import 'package:test/test.dart';
import 'package:universal_payment_qr/universal_payment_qr.dart';

// Points at a host build of crates/ffi (see bindings/flutter/README.md "Running tests") -
// production Android use resolves the bundled .so via NativeBindings' default Android branch
// instead; this override exists so desktop `dart test` can exercise the same FFI contract without
// a device/emulator.
const _libraryPathOverride =
    String.fromEnvironment('UPQR_NATIVE_LIB', defaultValue: '../../target/release/universal_payment_qr_ffi.dll');

void main() {
  test('parses UPI without floating point drift', () {
    final result = UniversalPaymentQr.parse(
      'upi://pay?pa=merchant%40bank&pn=Example&am=499.00&cu=INR',
      libraryPathOverride: _libraryPathOverride,
    );
    expect(result['recognized'], isTrue);
    expect(result['scheme'], 'upi');
    expect(result['amount'], '499.00');
  });

  test('recognizes a bare bitcoin address with no uri scheme', () {
    final result = UniversalPaymentQr.parse(
      '1BoatSLRHtKNngkdXEeobR76b53LETtpyT',
      libraryPathOverride: _libraryPathOverride,
    );
    expect(result['recognized'], isTrue);
    expect(result['scheme'], 'bitcoin');
  });

  test('non-payment QR is recognized but flagged, not left fully unrecognized', () {
    final result = UniversalPaymentQr.parse(
      'https://example.com/docs',
      libraryPathOverride: _libraryPathOverride,
    );
    expect(result['recognized'], isTrue);
    expect(result['scheme'], 'url');
    expect(result['supported'], isFalse);
    expect((result['support'] as Map)['reason'], 'NOT_PAYMENT_QR');
  });

  test('disabled scheme stays recognized but unsupported', () {
    final scanner = createScanner(
      schemes: {'solana_pay': false},
      libraryPathOverride: _libraryPathOverride,
    );
    final result = scanner.scan('solana:9xQeWvG816bUx9EPfEZi4q3G44E2sA6v7a6D9k5GmRse?amount=1.25');
    expect(result['recognized'], isTrue);
    expect(result['supported'], isFalse);
    expect((result['support'] as Map)['reason'], 'SCHEME_DISABLED');
  });

  test('preset scopes to india only', () {
    final scanner = createScanner(preset: 'india', libraryPathOverride: _libraryPathOverride);
    final upi = scanner.scan('upi://pay?pa=merchant%40bank&am=1&cu=INR');
    expect(upi['supported'], isTrue);
    final bitcoin = scanner.scan('1BoatSLRHtKNngkdXEeobR76b53LETtpyT');
    expect(bitcoin['recognized'], isTrue);
    expect(bitcoin['supported'], isFalse);
  });

  test('capabilities lists every registered scheme', () {
    final capabilities = UniversalPaymentQr.capabilities(libraryPathOverride: _libraryPathOverride);
    final ids = capabilities.map((s) => (s as Map)['id']).toSet();
    expect(ids, contains('upi'));
    expect(ids, contains('pix'));
    expect(capabilities.length, greaterThan(30));
  });

  test('supported schemes excludes experimental', () {
    final supported = UniversalPaymentQr.supportedSchemes(libraryPathOverride: _libraryPathOverride);
    for (final scheme in supported) {
      expect((scheme as Map)['maturity'], isNot('experimental'));
    }
  });

  test('version is non-empty', () {
    expect(UniversalPaymentQr.version(libraryPathOverride: _libraryPathOverride), isNotEmpty);
  });
}
